File: Python/Arrays/L_ImageSmoother.py
from typing import List


def imageSmoother(self, img: List[List[int]]) -> List[List[int]]:
    # Get the dimensions of the image matrix
    rows, cols = len(img), len(img[0])
    
    # Define a helper function to calculate the average value for a pixel
    def average_value(r, c):
        total, count = 0, 0

        # Define the boundaries for the neighboring pixels
        top = max(0, r - 1)
        bottom = min(rows, r + 2)
        left = max(0, c - 1)
        right = min(cols, c + 2)

        # Iterate over the neighboring pixels and calculate the sum and count
        for row in range(top, bottom):
            for col in range(left, right):
                total += img[row][col]
                count += 1

        # Calculate and return the average value for the pixel
        return total // count

    # Apply the average function to each pixel in the image matrix
    return [[average_value(r, c) for c in range(cols)] for r in range(rows)]

File: cpp/Array/L_ImageSmoother.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> imageSmoother(vector<vector<int>> &img)
{
    int rows = img.size();
    int cols = img[0].size();

    // Apply the average function to each pixel in the image matrix
    vector<vector<int>> result(rows, vector<int>(cols, 0));
    for (int r = 0; r < rows; r++)
    {
        for (int c = 0; c < cols; c++)
        {
            result[r][c] = calculateAverage(img, r, c, rows, cols);
        }
    }

    return result;
}

int calculateAverage(const vector<vector<int>> &img, int r, int c, int rows, int cols)
{
    int total = 0;
    int count = 0;

    // Define the boundaries for the neighboring pixels
    int top = max(0, r - 1);
    int bottom = min(rows, r + 2);
    int left = max(0, c - 1);
    int right = min(cols, c + 2);

    // Iterate over the neighboring pixels and calculate the sum and count
    for (int row = top; row < bottom; row++)
    {
        for (int col = left; col < right; col++)
        {
            total += img[row][col];
            count += 1;
        }
    }

    // Calculate and return the average value for the pixel
    return total / count;
}